Fix #3422: Separate MarkTopicsCompletedActivityTest into activity and fragment test files

yashraj-01 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fixes #3422. Separated MarkTopicsCompletedActivityTest into activity and fragment test files.

**Screenshots**

- Expresso test results for `MarkTopicsCompletedActivityTest`
![Screenshot from 2021-07-17 17-43-18](https://user-images.githubusercontent.com/63469151/126036775-a61832e7-9c2f-4b5e-9656-19a4c9236259.png)

- Expresso test results for `MarkTopicsCompletedFrgmentTest`
![Screenshot from 2021-07-17 17-26-03](https://user-images.githubusercontent.com/63469151/126036766-e7a60710-f015-4e89-a206-22ae5449d604.png)

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

yashraj-01 committed:
Fix #3422: Separate MarkTopicsCompletedActivityTest into activity and fragment test files

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkTopicsCompletedFragmentTest.kt:
```diff
@@ -0,0 +1,565 @@
+
+  @Test
+  fun testMarkTopicsCompletedFragment_topicsAreShown() {
+    launch<MarkTopicsCompletedActivity>(
```
If this test is launching MarkTopicsCompletedActivity then it's testing the activity, not the fragment. To test the fragment we should host it in a test activity on its own.

See https://github.com/oppia/oppia-android/blob/develop/app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt & https://github.com/oppia/oppia-android/blob/develop/app/src/main/java/org/oppia/android/app/player/state/testing/StateFragmentTestActivity.kt for an idea on how this works.

yashraj-01 replied:
I had a look at these and some other test activities and their corresponding prod files and I am a bit confused. 

Actually, I have noticed that test activities are there for cases where the fragment does not have its own activity, for example: `StateFragment` does not have a `StateActivity`.

Also, in some cases, we are launching the prod activity in `FragmentTest`, for example: `OnboardingFragmentTest` launches `OnboardingActivity` in its tests. There might be some reason for it that I may have missed.

Another case is where we have test activities even when the prod activities exist. Now if I'm to create a test activity in this case, I'm not exactly sure what functionalities should be there in this test activity. Also, in some of the cases, the test activity uses the prod activity presenter and in some cases, we have also created a test activity presenter. So, I'm also confused here.

Please help me understand this and guide me on how should I proceed. Thanks.

BenHenning replied:
These are good questions @yashraj-01. In general, components should always be tested in isolation (meaning that for any fragment it should always be tested via a test activity that hosts it). We aren't consistent on this everywhere today, and there are existing examples you pointed out that could also benefit from this change.

Regarding what the test activity should do: essentially the bare minimum to be able to test the fragment. We actually want to keep them as simple as possible, and often avoid introducing presenters altogether and just have the test activity.

Does this help clarify?

yashraj-01 replied:
Thanks @BenHenning for the clarification. I have introduced `MarkTopicsCompletedTestActivity` from what I have understood from your comments. PTAL and confirm if it is in the right direction or not? Thanks.

BenHenning replied:
Yep, this exactly addresses my comment. Thanks!

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkTopicsCompletedActivityTest.kt:
```diff
@@ -143,390 +105,12 @@ class MarkTopicsCompletedActivityTest {
     assertThat(title).isEqualTo(context.getString(R.string.mark_topics_completed_activity_title))
```
It might be worth having a few basic tests to verify that the correct UI is showing in the activity in both portrait & landscape. I don't think it pays to go into the same level of detail as the fragment tests, but something as simple has verifying the correct headers are being shown could be useful (consider that MarkTopicsCompletedActivity could never add the fragment & this test suite + the fragment test suite could still pass, assuming the latter is using a test activity as suggested in a later comment).

yashraj-01 replied:
Hmm. But the activity only has the `FrameLayout` and even the toolbar is part of the fragment. So, I'm not sure what additional tests we can add here. cc: @rt4914

yashraj-01 replied:
I have added a test (and its config change version) to test if the fragment is displayed or not. Will this be enough @BenHenning?

BenHenning replied:
Hmm. If you remove the code from MarkTopicsCompletedActivityPresenter that adds the fragment, do these tests still pass?

yashraj-01 replied:
No, the tests fail if I remove that code from `MarkTopicsCompletedActivityPresenter`.

BenHenning replied:
Just to double check: these specific tests in MarkTopicsCompletedActivityTest fail if that code is removed? If that's the case, then I think the tests you added here are sufficient.

yashraj-01 replied:
Yes.

BenHenning reviewed (commented):
Nice idea to ensure test coverage. Had a few comments--PTAL.

rt4914 commented:
Please add screenshot showing that Espresso Tests are passing for the files modified in this PR. Also once @BenHenning 's comments are resolved and fixed just ping me so that I can check it accordingly.

rt4914 commented:
Please add screenshot showing that Espresso Tests are passing for the files modified in this PR. Also once @BenHenning 's comments are resolved and fixed just ping me so that I can check it accordingly.

yashraj-01 committed:
Introduced MarkTopicsCompletedTestActivity

yashraj-01 committed:
Added test activity to exemptions

rt4914 reviewed (approved):
LGTM.

BenHenning commented on app/src/main/java/org/oppia/android/app/devoptions/marktopicscompleted/testing/MarkTopicsCompletedTestActivity.kt:
```diff
@@ -0,0 +1,46 @@
+import org.oppia.android.app.activity.InjectableAppCompatActivity
+import org.oppia.android.app.devoptions.marktopicscompleted.MarkTopicsCompletedFragment
+
+/** The activity for testing [MarkTopicsCompletedFragment] */
```
Nit: the summary fragment for classes should always end in punctuation.

yashraj-01 replied:
Added punctuation.

BenHenning commented on app/src/main/java/org/oppia/android/app/devoptions/marktopicscompleted/testing/MarkTopicsCompletedTestActivity.kt:
```diff
@@ -0,0 +1,46 @@
+    const val MARK_TOPICS_COMPLETED_TEST_ACTIVITY_PROFILE_ID_KEY =
+      "MarkTopicsCompletedTestActivity.internal_profile_id"
+
+    fun createMarkTopicsCompletedTestIntent(context: Context, internalProfileId: Int): Intent {
```
Please add a brief KDoc since this is a public API method.

yashraj-01 replied:
Done.

BenHenning reviewed (commented):
Thanks @yashraj-01! Left some more thoughts, but this in general looks good.

yashraj-01 committed:
Nit fixes

yashraj-01 commented:
@BenHenning Bazel build is successful locally but it is failing here. Is this an issue with the code or the workflow?
![Screenshot from 2021-07-20 14-58-28](https://user-images.githubusercontent.com/63469151/126298296-c19b1529-d363-47a4-985d-62779841b272.png)

BenHenning commented:
Hmm I'm not sure @yashraj-01. It might be a weird transient error or potentially a flake, but I've never seen the dex step fail specifically in CI before. I'm restarting the workflow to see if that helps.

In the meantime, I responded to the open comment thread above. PTAL & reassign to me once that thread is addressed & the latest CI flow has finished.

yashraj-01 commented:
> Hmm I'm not sure @yashraj-01. It might be a weird transient error or potentially a flake, but I've never seen the dex step fail specifically in CI before. I'm restarting the workflow to see if that helps.
> 
> In the meantime, I responded to the open comment thread above. PTAL & reassign to me once that thread is addressed & the latest CI flow has finished.

Actually, this issue is also faced by @ARJUPTA and @FareesHussain. So, I think there might be some issue with the CI.

yashraj-01 commented:
> Hmm I'm not sure @yashraj-01. It might be a weird transient error or potentially a flake, but I've never seen the dex step fail specifically in CI before. I'm restarting the workflow to see if that helps.
> 
> In the meantime, I responded to the open comment thread above. PTAL & reassign to me once that thread is addressed & the latest CI flow has finished.

@BenHenning I have addressed the comment. Also, the CI check is still failing with the same dex error.

BenHenning reviewed (approved):
Thanks @yashraj-01. LGTM since the CI failure should be fixed with #3513.

yashraj-01 committed:
Merge branch 'develop' into separate-mark-topics-completed-test

yashraj-01 commented:
Merged with latest develop. Also, assigning @Sarthak2601 for review as the mentor.

yashraj-01 committed:
Fixed failing espresso tests

Sarthak2601 reviewed (approved):
LGTM.
